### PR TITLE
[Feature] Atomic swap execution

### PR DESCRIPTION
**Pull request by ben-abraham**
Adds basic execution of atomic swaps directly into wallet.
Buy/Sell orders have been the most heavily tested so far.

image

### CODE REVIEW

File: src/qt/atomicswapsdialog.cpp [hunk 1]
```diff
@@ -0,0 +1,822 @@
...
+    
+    CTxIn swap_vin = result.Transaction.vin[0];
+    CTxOut swap_vout = result.Transaction.vout[0];
+
+    auto vin_script = ScriptToAsmStr(swap_vin.scriptSig, true);
+    auto fIsSingleSign = vin_script.find("[SINGLE|ANYONECANPAY]") != std::string::npos;
+
+    if(!fIsSingleSign)
+    {
+        errorMessage = tr("Must be signed with SINGLE|ANYONECANPAY");
```
Comment (fdoving) [inline src/qt/atomicswapsdialog.cpp:376]: Should probably split this, leave SINGLE|ANYONECANPAY out of tr().

File: src/wallet/wallet.h [hunk 2]
```diff
@@ -888,6 +879,15 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
...
+    /**
+     * Select a set of coins such that nValueRet >= nTargetValue and at least
+     * all coins from coinControl are selected; Never select unconfirmed coins
+     * if they are not ours
+     */
+    bool SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAmount& nTargetValue, std::set<CInputCoin>& setCoinsRet, CAmount& nValueRet, const CCoinControl *coinControl = nullptr) const;
+
+    bool SelectAssets(const std::map<std::string, std::vector<COutput> >& mapAvailableAssets, const std::map<std::string, CAmount>& mapAssetTargetValue, std::set<CInputCoin>& setCoinsRet, std::map<std::string, CAmount>& nValueRet) const;
+
     /**
```
Comment (fdoving) [inline src/wallet/wallet.h:891]: Why this 200 line move? - I can't see any changes.

Comment (ben-abraham) [reply src/wallet/wallet.h:891]: Moved from private to public.

File: src/qt/atomicswapsdialog.cpp (hunk 1 above)
Comment (ben-abraham) [reply src/qt/atomicswapsdialog.cpp:376]: Ok, what would you think replacement error should be? "Signed partial not constructed properly." or something like that?

File: src/wallet/wallet.h (hunk 2 above)
Comment (fdoving) [reply src/wallet/wallet.h:891]: github webpage camuflated that 👍

File: src/qt/atomicswapsdialog.cpp (hunk 1 above)
Comment (fdoving) [reply src/qt/atomicswapsdialog.cpp:376]: The message can be the same, but don't translate "SINGEL|ANYONECANPAY"

**Comment by HyperPeek** [comment]
Looks technically good to me. I see no alternative to make SelectCoins and SelectAssets public in wallet.h, as we definitely do not want to merge all the swap functionality into the wallet code.
Will do more testing first.

**Comment by ben-abraham** [comment]
> Looks technically good to me. I see no alternative to make SelectCoins and SelectAssets public in wallet.h, as we definitely do not want to merge all the swap functionality into the wallet code.
> Will do more testing first.

The internal working functions inside of SelectCoins and SelectAssets are already public as-is, so its not too strange to make the primary SelectCoins/SelectAssets public themselves.

File: src/qt/atomicswapsdialog.cpp (hunk 1 above)
Comment (ben-abraham) [reply src/qt/atomicswapsdialog.cpp]: resolved in 993f991

**Comment by hans-schmidt** [comment]
-This is a very interesting technology and a nice bit of coding work.
-Unfortunately, after discussing in the RavenQt-SIG multiple times, there is agreement at this time that this functionality does not belong in raven-qt
-This code provides support for completing (executing) pre-existing 2-step on-chain SIGHASH_SINGLE|SIGHASH_ANYONECANPAY atomic swap half-transactions. But creating those half-transactions still requires a separate application (the python app raven-trader-pro at this time). Even if creation of such half-transactions was also to raven-qt, there is no messaging chat protocol support for the nodes to share the available offer information.
-All of this adds too much complexity to the node, which can easily be supported by an external application.
-Including this would also generate a lot of user confusion and support questions.
-The basic ideas are extremely powerful and useful, but needed to be standardized, completed, and supported by an external app.

-I am closing this PR at this time. It can be re-opened later if community consensus justifies doing so.

**Pull request closed without merging**